[WIP] Interpreter boxing/unboxing

kg opened the pull request:
Adds interp opcodes for box, unbox, and unbox_any. The interpreter compiler looks up the appropriate helper ftn for the operation in question at compile time and then we invoke the helper to perform the actual box/unbox operation.

Also refactors a couple inconsistent names in the interpreter (MethodDesc/MethodToken -> MethodHandle)

kg committed:
Introduce box/unbox/unbox_any opcodes

kg committed:
Partially implement box opcode

kg committed:
Broken boxing implementation

kg committed:
Fix box helper call convention

kg committed:
Basic boxing/unboxing works

BrzVlad commented on src/coreclr/interpreter/intops.h:
```diff
@@ -21,6 +21,7 @@ typedef enum
     InterpOpBranch,
     InterpOpSwitch,
     InterpOpMethodToken,
+    InterpOpClassHandle,
```
Would be nice to have same unified suffix for these. Either `Token` or `Handle`

kg replied:
I propose Handle since the methodtoken one isn't actually a token.

kg replied:
I fixed this and also renamed METHOD_DESC_TAG to METHOD_HANDLE_TAG for consistency

am11 commented on src/tests/JIT/interpreter/Interpreter.cs:
```diff
@@ -386,4 +389,14 @@ public static bool TestVirtual()
+        if (j != i)
+            return false;
+        return true;
+    }
```
nit: from JIT viewpoint, this gets optimized to return true. We were recently using this benchmark for compulsory boxing:

```c#
using System.Runtime.CompilerServices;
...
    public object Box8() => Box(new Buffer8());
    
    [MethodImpl(MethodImplOptions.NoInlining | MethodImplOptions.NoOptimization)]
    static object Box<T>(T value) => (object)value;

    [InlineArray(1)]
    public struct Buffer8
    {
        long _element0;
    }
```

kg committed:
More robust test (mysteriously broken)

kg committed:
Fix box/unbox/unbox.any not popping the input from the interpreter stack
Fix unbox_any pushing VT unconditionally instead of pushing the appropriate interp type

kg committed:
Unify opcodes

kg committed:
Name normalization

kg committed:
Name normalization

kg committed:
Cleanup

janvorli commented on src/coreclr/interpreter/compiler.cpp:
```diff
@@ -3457,6 +3467,41 @@ int InterpCompiler::GenerateCode(CORINFO_METHOD_INFO* methodInfo)
+            }
+
+            case CEE_UNBOX:
+            case CEE_UNBOX_ANY: {
```
A nit - the convention used in this file has the opening brace on a separate line for the switch cases. The same for the CEE_BOX above.

janvorli commented on src/coreclr/interpreter/compiler.cpp:
```diff
@@ -3457,6 +3467,41 @@ int InterpCompiler::GenerateCode(CORINFO_METHOD_INFO* methodInfo)
+                CHECK_STACK(1);
+                m_pStackPointer -= 1;
+                CORINFO_CLASS_HANDLE clsHnd = ResolveClassToken(getU4LittleEndian(m_ip + 1)),
+                    boxedClsHnd = m_compHnd->getTypeForBox(clsHnd);
```
A nit - declaring two variables on a single line like this unusual for the coreclr codebase coding style.

janvorli reviewed (approved):
LGTM, thank you

kg committed:
Address PR feedback
